### PR TITLE
TEP0122 - Complete build instructions and parameters - implementable

### PR DESCRIPTION
**Pull request by chitrangpatel**
[This proposal](https://github.com/tektoncd/community/pull/820) was intended to be implementable but was accidentally left in proposed state. Some minor cleanup (accidental `buildConfig --> invocation.Parameters`) and clarifications were added to make it more readable. No major changes were introduced.

### CODE REVIEW

**Comment by afrittoli** [comment]
Maybe this PR is an opportunity to address https://github.com/tektoncd/community/pull/820#discussion_r1037520307?

**Comment by chitrangpatel** [comment]
> Maybe this PR is an opportunity to address [#820 (comment)](https://github.com/tektoncd/community/pull/820#discussion_r1037520307)?

Good point. I will move the conversation here:

>What about flags passed to the controller directly at start time?
>
Hmm. I didn't think of that. Can you provide an example of what these flags are? I'm used to running `ko apply -R config/`.

> Also, we have a lot of different configuration flags, so it would make sense to specify which ones shall be included.
> 
I was thinking of including them all since there are only [16](https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/apis/config/feature_flags.go#L97-L112) I think.
> It feels like this would cause a non-negligable increase in size of the TaskRun resources.
>
Hmm. Unless I'm misunderstanding, the only features that I'm thinking of are [these](https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/apis/config/feature_flags.go#L97-L112). Surely, that list could increase over time but probably not too large? That shouldn't be an issue right?
We can definitely update this approach if its not good enough. I've linked you to the original discussion with Billy as to why it would be nice to have the configuration information in the task run status.

Is it a standard SLSA requirement to include the configuration of the build system?
I think SLSA wants reproducibility as best as we can. I think feature flags that were used to configure Tekton that produced the build is very important for reproducibility.

**Comment by chuangw6** [comment]
Synced offline with @chitrangpatel . We need to update the tables to reflect that service account, workspaces etc. provided in taskrun will be recorded in invocation.params instead of buildConfig.

Not sure if it fits into this pr which just moves the tep to implementable. If not, feel free to do it in another pr :d Thanks @chitrangpatel !!

**Review by chuangw6 — COMMENTED** [review]
Thanks @chitrangpatel for updating this!

File: teps/0122-complete-build-instructions-and-parameters.md [hunk 1]
```diff
@@ -170,26 +170,26 @@ The table below provides the name and description of the API fields that the Tas
...
-| ComputeResources | Configure compute resources required by the steps in the task. | Required (because this could also indicate a minimum amount of resources required for the task run.) | buildConfig |
+| service account name | name of the service account | Required | invocation.Parameters |
+| params | parameter values provided by taskrun | Provided | invocation.Parameters |
+| workspaces | workspaces used by the task | Required | invocation.Parameters |
+| pod template | | Required | invocation.Parameters |
+| Timeout | time in which a task should complete | Required for completeness | invocation.Parameters |
+| StepOverrides | Override Step configuration specified in a Task. Currently we can override compute resources. | Required for completeness | invocation.Parameters |
+| SidecarOverrides | Override Sidecar configuration specified in a Task. Currently we can override compute resources. | Required for completeness | invocation.Parameters |
+| ComputeResources | Configure compute resources required by the steps in the task. | Required (because this could also indicate a minimum amount of resources required for the task run.) | invocation.Parameters |
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply teps/0122-complete-build-instructions-and-parameters.md:180]: @chuangw6 I agree. I meant to capture that structure in https://github.com/tektoncd/community/blob/main/teps/0122-complete-build-instructions-and-parameters.md#provenance-for-executed-taskrun 

Should I explain it in this section as well?

[the remark this answers is not in the document]
Comment (chitrangpatel) [reply teps/0122-complete-build-instructions-and-parameters.md:180]: I can update that section and put in the exact data types next to the names.

**Comment by chitrangpatel** [comment]
> Synced offline with @chitrangpatel . We need to update the tables to reflect that service account, workspaces etc. provided in taskrun will be recorded in invocation.params instead of buildConfig.
> 
> Not sure if it fits into this pr which just moves the tep to implementable. If not, feel free to do it in another pr :d Thanks @chitrangpatel !!

I think it fits here. Before making it implementable, its better to address these issues.

File: teps/0122-complete-build-instructions-and-parameters.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (chuangw6) [reply teps/0122-complete-build-instructions-and-parameters.md:180]: > @chuangw6 I agree. I meant to capture that structure in https://github.com/tektoncd/community/blob/main/teps/0122-complete-build-instructions-and-parameters.md#provenance-for-executed-taskrun
> Should I explain it in this section as well?

Oh that's great.  Yeah would be great to explain it a little bit perhaps just above the table?

On more thing: what is the first `name`?

```
    parameters:
      [taskRun](https://github.com/tektoncd/pipeline/blob/b301d88cf4af3301cfff40e3df6d85ad3848df0d/pkg/apis/pipeline/v1beta1/taskrun_types.go#L38-L86):
        name:
        serviceAccountName:
        computeResources:
        workspaces:
        podTemplate:
        taskRunResults:
        params:
        sidecarOverrides:
        stepOverrides:
        timeout:
```

[the remark this answers is not in the document]
Comment (chitrangpatel) [reply teps/0122-complete-build-instructions-and-parameters.md:180]: I did that for `TaskRun` under parameters where the underlying fields are defined.

[the remark this answers is not in the document]
Comment (chitrangpatel) [reply teps/0122-complete-build-instructions-and-parameters.md:180]: It was meant to be `taskRun` with a reference. But that did not come out well in this format 😢.

```
parameters:
  taskRun:
    name:
    ...
```

[the remark this answers is not in the document]
Comment (chitrangpatel) [reply teps/0122-complete-build-instructions-and-parameters.md:180]: Let me update it.

**Comment by chuangw6** [comment]
> What about flags passed to the controller directly at start time?
> Is it a standard SLSA requirement to include the configuration of the build system?

I think on the surface level, the provenance aims to record the information about the build process instead of the build system. There is indeed a field in the SLSA v0.2 provenance named [predicate.builder](https://slsa.dev/provenance/v0.2#builder) to record the builder information, but it only records the builder identifier.

That said, the SLSA v1.0 design ([draft](https://github.com/MarkLodato/slsa/blob/provenance-refactor/docs/provenance/v1.0.proto)) seems to introduce more fields about builder, and one of them is [builderDependencies](https://github.com/MarkLodato/slsa/blob/03a066000033fad54af5ce5d88f8aa672b130237/docs/provenance/v1.0.proto#L220-L226) which seems like a better place for the flags to the controller?

So I am leaning towards making flags out of scope for SLSA v0.2, but I am open to other suggestions here.

cc @afrittoli @chitrangpatel

**Comment by chitrangpatel** [comment]
@chuangw6 @afrittoli according to https://slsa.dev/provenance/v0.2, `invocation.environment` is meant to contain any builder-controlled inputs necessary for correctly evaluating the build. Usually only needed for [reproducing](https://reproducible-builds.org/) the build but not evaluated as part of policy. 

In slsav0.2, I think this is where config-feature-flags belong and may be also arguments that I think @afrittoli is mentioning.

**Comment by pritidesai** [comment]
API WG - @afrittoli please take a look, this approved by other reviewers. @chitrangpatel to rebase this. Thanks!

**Review by afrittoli — COMMENTED** [review]
/approve

File: teps/README.md [hunk 2]
```diff
@@ -288,8 +288,8 @@ This is the complete list of Tekton teps:
 |[TEP-0118](0118-matrix-with-explicit-combinations-of-parameters.md) | Matrix with Explicit Combinations of Parameters | implementable | 2022-08-08 |
 |[TEP-0119](0119-add-taskrun-template-in-pipelinerun.md) | Add taskRun template in PipelineRun | implementable | 2022-09-01 |
 |[TEP-0120](0120-canceling-concurrent-pipelineruns.md) | Canceling Concurrent PipelineRuns | proposed | 2022-09-23 |
-|[TEP-0121](0121-refine-retries-for-taskruns-and-customruns.md) | Refine Retries for TaskRuns and CustomRuns | implemented | 2022-12-21 |
-|[TEP-0122](0122-complete-build-instructions-and-parameters.md) | Complete Build Instructions and Parameters | proposed | 2022-11-29 |
+|[TEP-0121](0121-refine-retries-for-taskruns-and-customruns.md) | Refine Retries for TaskRuns and CustomRuns | implementable | 2022-12-01 |
```
Comment (bendory) [inline teps/README.md:291]: I assume you didn't intend to change this line?

Comment (chitrangpatel) [reply teps/README.md:291]: Yes. Thanks for that catch. Merge conflict issues :(

Comment (chitrangpatel) [reply teps/README.md]: fixed now.

**Comment by afrittoli** [comment]
Approved in API WG
/lgtm

**Pull request merged**